pkg/scheduler/framework/preemption/preemption.go

```diff
@@ -573,6 +573,7 @@ func (ev *Evaluator) DryRunPreemption(ctx context.Context, pod *v1.Pod, potentia
 	nonViolatingCandidates := newCandidateList(numCandidates)
 	violatingCandidates := newCandidateList(numCandidates)
 	parallelCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
```

Comment:
I'm trying to understand why it could reduce the potential goroutine leak:

- if parallelCtx gets closed by an early cancel() (i.e., in one of the `checkNode` goroutine). `cancel()` here is redundant as it's a double cancellation.
- if all `checkNode` goroutines finish successfully. `cancel()` here is also sort of redundant as if all nodes get looped, it's not necessary to call cancel().

So I'm skeptical if cancel() is the right way to fix the issue unless I miss anything.

Could you try to keep this line and change the `ctx` in L583 to `parallelCtx`? which makes more sense to me and I guess it can also fix the issue.

Reply:
In my understanding, we should call cancel in this case as well.
Because parent context has the context created here. And parent ctx won't cancel them until he is canceled. 

> all nodes get looped, it's not necessary to call cancel().

Reply:
>parent ctx won't cancel them until he is canceled.

And parent ctx won't be canceled until scheduler stops

Reply:
> Because parent context has child contexts created here internally.

If all nodes get looped, not canceling the child ctx is not a big deal - it will be GCed. The exceptional case that goroutines leak inside `checkNode()`. I checked its logic but didn't find any suspicious place it could leak. Could you help take a second look?

Taking a step back, even if we have goroutine leaked in `checkNode()`, in current code, it's passing the **parent** context, so canceling **child** context won't help either. That's why i suggested changing the logic ^^ https://github.com/kubernetes/kubernetes/pull/111773#discussion_r943066878.

Reply:
They are not GCed.

> Calling the CancelFunc cancels the child and its children, removes the parent's reference to the child, and stops any associated timers. Failing to call the CancelFunc leaks the child and its children until the parent is canceled or the timer fires.
https://pkg.go.dev/context

Reply:
> They are not GCed.
> 
> > Calling the CancelFunc cancels the child and its children, removes the parent's reference to the child, and stops any associated timers. Failing to call the CancelFunc leaks the child and its children until the parent is canceled or the timer fires.
> > https://pkg.go.dev/context

I agree with you. Heap profile generated by `pprof` is able to prove your thought. In my opinion, current logic might cover all the cases that finally close `parallelCtx`. Calling `CancelFunc` here could  guarantee that `parallelCtx` will be GCed after `DryRunPreemption` finished. As @Huang-Wei said, it's sort of redundant calling though.

Reply:
Thanks for the reference to the docs.
I just double checked and it seems that the parent context is the context of the entire scheduler worker. So it's definitely leaking 😞 

I also noticed that we create a context just for a scheduling cycle in `scheduler.go#scheduleOne`. However, we are not using this context for `PostBind`. We should, to prevent any other accidental leaks. Could you fix that too?

TBH, I would do something like:

```
ctx, cancel = context.WithCancel(ctx)
```

right at the top of the `scheduleOne` func.

Reply:
👍
That will be fixed in https://github.com/kubernetes/kubernetes/pull/111775

Reply:
> TBH, I would do something like:
> 
> ```
> ctx, cancel = context.WithCancel(ctx)
> ```
> 
> right at the top of the `scheduleOne` func.

Actually, don't do this, because we need a separate context for the binding cycle. Just use the `schedulingCycleCtx` for `RunPostFilterPlugins`.

Reply:
Ah #111775 is doing this.

/lgtm
/approve

Reply:
For posterity, here is the relevant quote from the docs:

```
Failing to call the CancelFunc leaks the child and its children until the parent is canceled or the timer fires. The go vet tool checks that CancelFuncs are used on all control-flow paths.
```

I am wondering if we run go vet on k8s code? and if so, why this one wasn't caught.

Reply:
/hold
Maybe we still need to update the L583. Replace the ctx with parallelCtx.

Reply:
That's tangential. Those functions would finish. But feel free to open a separate PR.

/hold cancel

Reply:
Also, I think that context is not being used. It's just passed around to the plugins. I don't think any of them use it. I could be wrong, but I'm pretty sure that even if they use the context, it's not a problem.

Reply:
But in general, it seems safer to always call `defer cancel()`.

What we could do is perhaps make that change to `ParallelizeUntil` to create the context inside it from the passed context, and change the function signature (like checkNode) to accept a context, but that also seem error prone.

Reply:
> Also, I think that context is not being used. It's just passed around to the plugins. I don't think any of them use it. I could be wrong, but I'm pretty sure that even if they use the context, it's not a problem.

I'll have a look when I have time.

Reply:
> What we could do is perhaps make that change to ParallelizeUntil to create the context inside it from the passed context

But we need to make sure that anything passing a context to `ParallelizeUntil` is not creating a subcontext before calling the function too. Otherwise it would still leak.